boot/bootstate20.go

```diff
@@ -410,61 +415,51 @@ type bootState20MarkSuccessful struct {
 	bootState20Kernel
 }
 
-// threadBootState20MarkSuccessful is a helper method that will either create a
-// new bootState20MarkSuccessful for the given type, or it will add to the
-// provided bootStateUpdate
-func threadBootState20MarkSuccessful(update bootStateUpdate) (*bootState20MarkSuccessful, error) {
-	var bsmark *bootState20MarkSuccessful
+// chooseBootSnapToMarkSuccessful inspects the specified boot state to pick what
+// boot snap should be marked as successful and use as a valid rollback target.
+// If the
+func chooseBootSnapToMarkSuccessful(b bootState, update bootStateUpdate) (
+	bsmark *bootState20MarkSuccessful,
+	bootedSnap snap.PlaceInfo,
+	err error,
+) {
+	// get the try snap and the current status
+	sn, trySnap, status, err := b.revisions()
+	if err != nil {
+		return nil, nil, err
+	}
 
 	// try to extract bsmark out of update
 	var ok bool
 	if update != nil {
 		if bsmark, ok = update.(*bootState20MarkSuccessful); !ok {
-			return nil, fmt.Errorf("internal error, cannot thread %T with update for UC20", update)
+			return nil, nil, fmt.Errorf("internal error, cannot thread %T with update for UC20", update)
 		}
 	}
 
 	if bsmark == nil {
 		bsmark = &bootState20MarkSuccessful{}
 	}
 
-	// initialize both types in case we need to mark both
-	err := bsmark.loadBootenv()
-	if err != nil {
-		return nil, err
-	}
-	err = bsmark.loadModeenv()
-	if err != nil {
-		return nil, err
-	}
-
-	return bsmark, nil
-}
-
-// genericMarkSuccessful sets the necessary boot variables, etc. to mark the
-// given boot snap as successful and a valid rollback target. If err is nil,
-// then the first return value is guaranteed to always be non-nil.
-func genericMarkSuccessful(b bootState, update bootStateUpdate) (bsmark *bootState20MarkSuccessful, trySnap snap.PlaceInfo, err error) {
-	// create a new object or combine the existing one with this type
-	bsmark, err = threadBootState20MarkSuccessful(update)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	// get the try snap and the current status
-	_, trySnap, status, err := b.revisions()
-	if err != nil {
-		return nil, nil, err
+	// incorporate the bootState into the bsmark so we don't need to reload
+	// bootenv / modeenv, as bootState will already have been initialized in
+	// the call to revisions() above
```

Comment:
I don't have an immediate suggestion but the switch below feels a bit awkward

Reply:
can't simply the callers do this instead ?  as long as the doc comment for select(choose)BootSnapToMarkSuccessful  tells that it loads b bootState with env state

Reply:
yes it's more code actually and requires re-introducing the thread helper, but then we can avoid the switch I guess

Reply:
I'm not sure why you think it needs the thread helper? I'm proposing to do it after the select* call , not before

Reply:
yes I got it now, thanks for the suggestion